HFnose: add L1 geometry and initial trigger cells

mariadalfonso opened the pull request:
Changes in the  L1Trigger/L1THGCal to introduce the TP for the HFnose.

Initial set of results here:
https://indico.cern.ch/event/855644/#71-triggering-with-the-hfnose

To run here is  
cmsRun L1Trigger/L1THGCal/test/testHGCalL1TGeometryV9Imp2_Nose_cfg.py

mariadalfonso committed:
add L1 geometry and initial trigger cells

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalClusterT.h:
```diff
@@ -114,14 +115,18 @@ namespace l1t {
+	    (id.det() == DetId::HGCalTrigger && HGCalTriggerDetId(id_constituent.first).subdet() == HGCalTriggerSubdetector::HGCalEETrigger) ||
+	    (id.det() == DetId::Forward && id.subdetId() == ForwardSubdetector::HFNose && HFNoseDetId(id_constituent.first).isEE()) ||
+	    (id.det() == DetId::HGCalTrigger && HGCalTriggerDetId(id_constituent.first).subdet() == HGCalTriggerSubdetector::HFNoseTrigger && HFNoseTriggerDetId(id_constituent.first).isEE())
+	    ) {
```
- Mixed indents to be fixed
- `HGCalTriggerDetId(id)` (and similar) can be used instead of `HGCalTriggerDetId(id_constituent.first)`. More readable.

mariadalfonso replied:
this lead to the code crash

jbsauvan replied:
I'm not sure to understand why it would crash. Is it in the `HFNoseTriggerDetId` constructor? There is an existing constructor from `DetId`, so if it crashes, there is something to be fixed elsewhere, but it is not a reason to change the code here.

mariadalfonso replied:
The original HGCalTriggerDetId( DetId(id_constituent.first) )  crash when a nose objects comes; HGCalTriggerDetId( id_constituent.first ) works because the rawId is passed directly.
I discussed with Sunanda once.

I apply in HGCalClusterT.h  what I did also in the in other parts, i.e. also in 
L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc

jbsauvan replied:
Do we know why it is crashing? I would prefer that it is fixed upstream.

mariadalfonso replied:
done, needs https://github.com/cms-sw/cmssw/pull/28249

jbsauvan replied:
I backported the fix in #404

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalClusterT.h:
```diff
@@ -114,14 +115,18 @@ namespace l1t {
+		   (id.det() == DetId::HGCalTrigger && HGCalTriggerDetId(id_constituent.first).subdet() == HGCalTriggerSubdetector::HGCalHSiTrigger) ||
+		   (id.det() == DetId::Forward && id.subdetId() == ForwardSubdetector::HFNose && HFNoseDetId(id_constituent.first).isHE()) ||
+		   (id.det() == DetId::HGCalTrigger && HGCalTriggerDetId(id_constituent.first).subdet() == HGCalTriggerSubdetector::HFNoseTrigger && HFNoseTriggerDetId(id_constituent.first).isHSilicon())
+		   ) {
```
Ditto

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalTriggerCellCalibration.h:
```diff
@@ -2,31 +2,34 @@
   double LSB_silicon_fC_;
   double LSB_scintillator_MIP_;
   double fCperMIP_;
+  double fCperMIPnose_;
```
Is there a reason that the fC per MIP could be different for the HF nose? I thought the same modules were used.

mariadalfonso replied:
All these parameters change on the geometry (HFnose will stay with the V9 for a while). I prefer to be independent on these calibration inputs. I think we can adjust later.

jbsauvan replied:
ok

jbsauvan commented on L1Trigger/L1THGCal/interface/backend/HGCalHistoSeedingImpl.h:
```diff
@@ -159,7 +159,10 @@ class HGCalHistoSeedingImpl {
+  // HGCAL value
+  //  static constexpr double kROverZMin_ = 0.076;
+  // HGCAL nose
+  static constexpr double kROverZMin_ = 0.02;  // r=30 cm at z=1000 cm
```
The number of bins has to be changed accordingly to keep the same bin size as before

mariadalfonso replied:
ok, will adjust and check with the other suggestion

mariadalfonso replied:
@jbsauvan Do you have a description on how change the binning.
doesn't seems a 1:1 scaling with the range

jbsauvan replied:
Actually it is. The bin size is constant in R/Z. The number of bins can be increased by the ratio of the new vs old boundaries (rounded to the closest integer). R/Z bins are given in `nBins1_` (there are generic names because we can have different types of binnings). The binning should both be modified in the source default value and in the python config.

mariadalfonso replied:
done, now we have 46 bins

jbsauvan commented:
There are **a lot** of useless formatting changes and changes in includes order. 
The formatting is following the rules defined in `scram b code-format`. Please run this command and push the proposed changes.

jbsauvan reviewed (commented):
First (incomplete) set of comments. I'll continue once the formatting update is done.

mariadalfonso commented:
@jbsauvan 
I pushed (and squashed) the code format changes.

jbsauvan commented on Fireworks/Calo/plugins/FWL1THGCalProxyTemplate.cc:
```diff
@@ -69,4 +71,4 @@ class FWL1THGCalProxyTemplate : public FWSimpleProxyBuilderTemplate<T> {
   }
-};
\ No newline at end of file
+};
```
It looks like there is no real update in this file. Are these changes needed?

mariadalfonso replied:
the #include "DataFormats/ForwardDetId/interface/HGCalTriggerDetId.h" 
was added

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalTriggerCellCalibration.h:
```diff
@@ -2,12 +2,12 @@
-#include "FWCore/ParameterSet/interface/ParameterSet.h"
-#include "FWCore/Framework/interface/ESHandle.h"
 #include "DataFormats/Common/interface/Handle.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "Geometry/HGCalCommonData/interface/HGCalDDDConstants.h"
```
No reason to change the include order

mariadalfonso replied:
the reordering were introduced by the code-format

jbsauvan replied:
Ok. That's strange. I have been applying it with the original ordering and it was not changing anything.

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h:
```diff
@@ -2,16 +2,16 @@
 #include <unordered_map>
+#include <unordered_set>
 
 #include "DataFormats/GeometryVector/interface/GlobalPoint.h"
 
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 
-#include "Geometry/HGCalGeometry/interface/HGCalGeometry.h"
-#include "Geometry/CaloTopology/interface/HGCalTopology.h"
 #include "Geometry/CaloGeometry/interface/CaloGeometry.h"
+#include "Geometry/CaloTopology/interface/HGCalTopology.h"
+#include "Geometry/HGCalGeometry/interface/HGCalGeometry.h"
```
No reason to change the include order

mariadalfonso replied:
the re-order was introduced by one of the code-format command.
(It seems there is a logic with the alphabet).

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalTriggerTools.h:
```diff
@@ -5,9 +5,9 @@
- *  when used for offline reco!!!!
+ *  As a consequence the class assumes that only DetIds of the first kind are
+ * used in the getTC* methods NOTE: this uses the trigger geometry hence would
+ * give wrong results when used for offline reco!!!!
```
No reason to change this

mariadalfonso replied:
This and other lines return were introduced by one of the code-format command.

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalTriggerTools.h:
```diff
@@ -18,8 +18,8 @@
 
 #include "DataFormats/DetId/interface/DetId.h"
-#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
 #include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
```
No reason to change the include order

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalTriggerTools.h:
```diff
@@ -72,7 +73,8 @@ class HGCalTriggerTools {
     std::vector<T> outputVector;
-    //loop over collection for a given bx and put the objects into a std::vector
+    // loop over collection for a given bx and put the objects into a
+    // std::vector
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/interface/backend/HGCalHistoSeedingImpl.h:
```diff
@@ -1,14 +1,14 @@
 
-#include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "DataFormats/L1THGCal/interface/HGCalCluster.h"
 #include "DataFormats/L1THGCal/interface/HGCalMulticluster.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
 
 #include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h"
-#include "L1Trigger/L1THGCal/interface/backend/HGCalShowerShape.h"
 #include "L1Trigger/L1THGCal/interface/HGCalTriggerTools.h"
+#include "L1Trigger/L1THGCal/interface/backend/HGCalShowerShape.h"
```
No reason to change the include order

jbsauvan commented on L1Trigger/L1THGCal/plugins/HGCalTriggerGeometryESProducer.cc:
```diff
@@ -1,9 +1,9 @@
 #include <memory>
 
-#include "FWCore/Framework/interface/ModuleFactory.h"
 #include "FWCore/Framework/interface/ESProducer.h"
 #include "FWCore/Framework/interface/ESProducts.h"
+#include "FWCore/Framework/interface/ModuleFactory.h"
```
No reason to change the include order

jbsauvan commented on L1Trigger/L1THGCal/plugins/HGCalTriggerGeometryESProducer.cc:
```diff
@@ -48,25 +50,27 @@ HGCalTriggerGeometryESProducer::~HGCalTriggerGeometryESProducer() {
 
 HGCalTriggerGeometryESProducer::ReturnType HGCalTriggerGeometryESProducer::produce(const CaloGeometryRecord& iRecord) {
-  //using namespace edm::es;
+  // using namespace edm::es;
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/HGCalTriggerGeometryESProducer.cc:
```diff
@@ -48,25 +50,27 @@ HGCalTriggerGeometryESProducer::~HGCalTriggerGeometryESProducer() {
 }
 
-//define this as a plug-in
+// define this as a plug-in
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/HGCalVFEProducer.cc:
```diff
@@ -1,13 +1,13 @@
 #include "FWCore/Framework/interface/Frameworkfwd.h"
-#include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
 
 #include "FWCore/Framework/interface/Event.h"
 #include "FWCore/Framework/interface/MakerMacros.h"
 
+#include "DataFormats/HGCDigi/interface/HGCDigiCollections.h"
 #include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
 #include "DataFormats/L1THGCal/interface/HGCalTriggerSums.h"
-#include "DataFormats/HGCDigi/interface/HGCDigiCollections.h"
```
No reason to change the include order

jbsauvan commented on L1Trigger/L1THGCal/plugins/HGCalVFEProducer.cc:
```diff
@@ -62,19 +63,24 @@ void HGCalVFEProducer::produce(edm::Event& e, const edm::EventSetup& es) {
 
-  // Processing DigiCollections and putting the results into the HGCalTriggerCellBxCollection
+  // Processing DigiCollections and putting the results into the
+  // HGCalTriggerCellBxCollection
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryHexImp1.cc:
```diff
@@ -1,11 +1,11 @@
-#include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryGenericMapping.h"
 #include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryGenericMapping.h"
 
-#include <vector>
-#include <iostream>
 #include <fstream>
+#include <iostream>
+#include <vector>
```
No reason to change the include order

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryHexImp1.cc:
```diff
@@ -105,9 +117,10 @@ void HGCalTriggerGeometryHexImp1::fillMaps()
+  // For each wafer compute the trigger cell offset according to the wafer
+  // position inside the module. The first wafer will have an offset equal to 0,
+  // the second an offset equal to the number of trigger cells in the first
+  // wafer, etc.
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryHexImp1.cc:
```diff
@@ -145,10 +158,12 @@ void HGCalTriggerGeometryHexImp1::fillMaps()
       // The module id is used instead of the wafer id for the trigger cells
-      // Since there are several wafers per module, an offset is applied on the HGCalDetId::cell field
+      // Since there are several wafers per module, an offset is applied on the
+      // HGCalDetId::cell field
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryHexImp1.cc:
```diff
@@ -185,10 +200,12 @@ void HGCalTriggerGeometryHexImp1::fillMaps()
       // The module id is used instead of the wafer id for the trigger cells
-      // Since there are several wafers per module, an offset is applied on the HGCalDetId::cell field
+      // Since there are several wafers per module, an offset is applied on the
+      // HGCalDetId::cell field
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryHexImp1.cc:
```diff
@@ -223,7 +240,7 @@ void HGCalTriggerGeometryHexImp1::buildTriggerCellsAndModules()
     auto itr_exist = trigger_cells_to_cells.emplace(triggerCell, list_cells());
     itr_exist.first->second.emplace(cell);
-    //trigger_cells_to_cells.at(triggerCell).emplace(cell);
+    // trigger_cells_to_cells.at(triggerCell).emplace(cell);
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryHexImp1.cc:
```diff
@@ -241,7 +258,7 @@ void HGCalTriggerGeometryHexImp1::buildTriggerCellsAndModules()
                              ? triggerCellToModuleItr->second
                              : 0);  // 0 if the trigger cell doesn't belong to a module
-    //unsigned moduleId = trigger_cells_to_modules_.at(triggercellId);
+    // unsigned moduleId = trigger_cells_to_modules_.at(triggercellId);
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryHexImp1.cc:
```diff
@@ -258,7 +275,7 @@ void HGCalTriggerGeometryHexImp1::buildTriggerCellsAndModules()
     auto itr_exist = modules_to_trigger_cells.emplace(module, list_triggerCells());
     itr_exist.first->second.emplace(triggerCell);
-    //modules_to_trigger_cells.at(module).emplace(triggerCell);
+    // modules_to_trigger_cells.at(module).emplace(triggerCell);
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryHexImp2.cc:
```diff
@@ -1,20 +1,21 @@
-#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
 #include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h"
 
-#include <vector>
-#include <iostream>
 #include <fstream>
+#include <iostream>
 #include <regex>
+#include <vector>
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryHexLayerBasedImp1.cc:
```diff
@@ -1,21 +1,22 @@
+#include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
 #include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
 #include "DataFormats/HcalDetId/interface/HcalDetId.h"
-#include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h"
 
-#include <vector>
-#include <iostream>
 #include <fstream>
+#include <iostream>
 #include <regex>
+#include <vector>
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryImp1.cc:
```diff
@@ -1,19 +1,20 @@
-#include "DataFormats/ForwardDetId/interface/HGCTriggerDetId.h"
 #include "DataFormats/ForwardDetId/interface/HGCEEDetId.h"
+#include "DataFormats/ForwardDetId/interface/HGCTriggerDetId.h"
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryGenericMapping.h"
 
-#include <vector>
-#include <iostream>
 #include <fstream>
+#include <iostream>
+#include <vector>
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryImp1.cc:
```diff
@@ -116,7 +128,7 @@ void HGCalTriggerGeometryImp1::buildMaps()
         (tc2mItr != trigger_cells_to_modules_.end() ? tc2mItr->second
                                                     : 0);  // 0 if the trigger cell doesn't belong to a module
-    //unsigned moduleId = trigger_cells_to_modules_.at(triggercellId);
+    // unsigned moduleId = trigger_cells_to_modules_.at(triggercellId);
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryImp1.cc:
```diff
@@ -138,7 +150,8 @@ void HGCalTriggerGeometryImp1::buildMaps()
     tc_map_to_cells cellsInTriggerCells;
-    // Position: for the moment, barycenter of the module, from trigger cell positions
+    // Position: for the moment, barycenter of the module, from trigger cell
+    // positions
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp1.cc:
```diff
@@ -1,21 +1,23 @@
+#include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
 #include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
 #include "DataFormats/HcalDetId/interface/HcalDetId.h"
-#include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h"
 
-#include <vector>
-#include <iostream>
 #include <fstream>
+#include <iostream>
 #include <regex>
+#include <vector>
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp1.cc:
```diff
@@ -171,8 +173,39 @@ void HGCalTriggerGeometryV9Imp1::initialize(const HGCalGeometry* hgc_ee_geometry
+  fillNeighborMap(l1tCellNeighborsMapping_, trigger_cell_neighbors_,
+                  false);  // silicon
+  fillNeighborMap(l1tCellNeighborsSciMapping_, trigger_cell_neighbors_sci_,
+                  true);  // scintillator
```
It seems that the HFNose is not implemented here, so shouldn't there be an exception?

mariadalfonso replied:
ok will add one, it's just not used 
I was only focusing on the HGCalTriggerGeometryV9Imp2.cc 
What those are are meant for ?

jbsauvan replied:
`V9Imp1` was the first implementation of the trigger geometry compatible with >=V9 geometries. It is not used by default anymore but could eventually be used.

mariadalfonso replied:
I will implement if needed beyond the 11_0 is out; add a note in the text

mariadalfonso replied:
implemented Expection on V1

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp1.cc:
```diff
@@ -243,7 +278,8 @@ unsigned HGCalTriggerGeometryV9Imp1::getModuleFromTriggerCell(const unsigned tri
   if (trigger_cell_det_id.subdetId() == ForwardSubdetector::HGCHEB) {
-    // For scintillator, the module ID is currently encoded as the wafer in HGCalDetId
+    // For scintillator, the module ID is currently encoded as the wafer in
+    // HGCalDetId
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -1,24 +1,26 @@
-#include "DataFormats/ForwardDetId/interface/HGCScintillatorDetId.h"
 #include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
+#include "DataFormats/ForwardDetId/interface/HFNoseTriggerDetId.h"
+#include "DataFormats/ForwardDetId/interface/HGCScintillatorDetId.h"
 #include "DataFormats/ForwardDetId/interface/HGCSiliconDetIdToROC.h"
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "DataFormats/ForwardDetId/interface/HGCalTriggerDetId.h"
+#include "DataFormats/HcalDetId/interface/HcalDetId.h"
+#include "FWCore/ParameterSet/interface/FileInPath.h"
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h"
+#include "DataFormats/ForwardDetId/interface/HFNoseDetIdToModule.h"
 
-#include <vector>
-#include <iostream>
 #include <fstream>
+#include <iostream>
 #include <regex>
+#include <vector>
```
Include order to be kept as it was

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -67,8 +69,9 @@ class HGCalTriggerGeometryV9Imp2 : public HGCalTriggerGeometryBase {
-  unsigned totalLayers_ = 0;
+  unsigned heOffset_;
+  unsigned noseLayers_;
+  unsigned totalLayers_;
```
The default values should be added back

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -135,6 +138,36 @@ void HGCalTriggerGeometryV9Imp2::initialize(const HGCalGeometry* hgc_ee_geometry
+  heOffset_ = eeTopology().dddConstants().layers(true);
+  totalLayers_ = heOffset_ + hsiTopology().dddConstants().layers(true);
+  noseLayers_ = noseTopology().dddConstants().layers(true);
+  totalLayers_ += noseLayers_;
```
The total number of layers is the total number of layers in HGCAL. I suppose it should be kept independent of the nose layers. I'm not sure that how it is done here doesn't have impact elsewhere given that it changes the mapping between detector layers and trigger layers. 
How are the nose layers numbered?

mariadalfonso replied:
For the nose we just need to have from 0 to 7.
I will keep invariant the totalLayers_ and add a totalNoseLayers_

mariadalfonso replied:
done, we have no a separate counter/handling for the noseTriggerLayers ( even if are 1:1 with the layer)

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -144,20 +177,34 @@ unsigned HGCalTriggerGeometryV9Imp2::getTriggerCellFromCell(const unsigned cell_
     int ieta = ((cell_sc_id.ietaAbs() - 1) / hSc_triggercell_size_ + 1) * cell_sc_id.zside();
     int iphi = (cell_sc_id.iphi() - 1) / hSc_triggercell_size_ + 1;
-    trigger_cell_id = HGCScintillatorDetId(cell_sc_id.type(), cell_sc_id.layer(), ieta, iphi);
+    trigger_cell_id = HGCScintillatorDetId(cell_sc_id.type(), cell_sc_id.layer(), ieta, iphi).rawId();
```
The implicit conversion from DetId to unsigned should be working. (same thing below)

mariadalfonso replied:
removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -295,6 +374,16 @@ HGCalTriggerGeometryBase::geom_set HGCalTriggerGeometryV9Imp2::getTriggerCellsFr
+    HFNoseDetId module_nose_id(module_id);
+    HFNoseDetIdToModule hfn;
+    std::vector<HFNoseTriggerDetId> ids = hfn.getTriggerDetIds(module_nose_id);
+    for (unsigned int tdi = 0; tdi < ids.size(); tdi++) {
```
Would be more readable with a range-based loop in that case

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -348,6 +438,16 @@ HGCalTriggerGeometryBase::geom_ordered_set HGCalTriggerGeometryV9Imp2::getOrdere
+    HFNoseDetId module_nose_id(module_id);
+    HFNoseDetIdToModule hfn;
+    std::vector<HFNoseTriggerDetId> ids = hfn.getTriggerDetIds(module_nose_id);
+    for (unsigned int tdi = 0; tdi < ids.size(); tdi++) {
```
Would be more readable with a range-based loop

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -393,6 +493,10 @@ unsigned HGCalTriggerGeometryV9Imp2::getLinksInModule(const unsigned module_id)
     links = hSc_links_per_module_;
   }
+  if (module_det_id.det() == DetId::Forward && module_det_id.subdetId() == ForwardSubdetector::HFNose) {
+    links = 1;
```
It should not be hardcoded. In particular if we use the "decentralized" ECON architecture (which is now the baseline), the number of links (elinks in that case) should be at least 2.

mariadalfonso replied:
yes, the links are the missing features.
to be added in the future, outside of this PR.

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -442,7 +555,7 @@ GlobalPoint HGCalTriggerGeometryV9Imp2::getTriggerCellPosition(const unsigned tr
     }
   }
-  return GlobalPoint(triggerCellVector / cell_ids.size());
+  return GlobalPoint(triggerCellVector / static_cast<int>(cell_ids.size()));
```
I'm not sure to see why a cast to int is needed

jbsauvan replied:
It's still unclear to me why a cast to signed integer has been added here.

mariadalfonso replied:
i.e. this casting is  also done here https://github.com/cms-sw/cmssw/pull/27874/files#diff-2579cb69237e24fcec0801b8586ea151

jbsauvan replied:
Still, I don't understand why something which was working has been changed here. Besides, a vector size is unsigned so I don't see the logic of casting it to a signed integer. 
If there is valid reason for this change, I'm ok with that (but I'd like to know it); if there is no valid reason, please keep this code (which is independent from the HFNose updates) unchanged.

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -465,7 +584,8 @@ GlobalPoint HGCalTriggerGeometryV9Imp2::getModulePosition(const unsigned module_
   }
-  return GlobalPoint(moduleVector / cell_ids.size());
+
+  return GlobalPoint(moduleVector / static_cast<int>(cell_ids.size()));
```
Ditto

jbsauvan replied:
Same thing

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -474,6 +594,9 @@ void HGCalTriggerGeometryV9Imp2::fillMaps() {
     throw cms::Exception("MissingDataFile") << "Cannot open HGCalTriggerGeometry L1TModulesMapping file\n";
   }
+
+  std::cout << l1tModulesMapping_.fullPath() << std::endl;
```
`cout` to be removed

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -628,8 +754,10 @@ int HGCalTriggerGeometryV9Imp2::detIdWaferType(unsigned det, unsigned layer, sho
 }
 
 unsigned HGCalTriggerGeometryV9Imp2::layerWithOffset(unsigned id) const {
+  // not used for HFNose
```
This function should be able to handle HFNose, otherwise it would always return 0 in that case?

mariadalfonso replied:
implemented

jbsauvan commented on L1Trigger/L1THGCal/python/hgcalVFEProducer_cfi.py:
```diff
@@ -72,6 +81,7 @@
         eeDigis = cms.InputTag('simHGCalUnsuppressedDigis:EE'),
         fhDigis = cms.InputTag('simHGCalUnsuppressedDigis:HEfront'),
         bhDigis = cms.InputTag('simHGCalUnsuppressedDigis:HEback'),
+        noseDigis = cms.InputTag('simHFNoseUnsuppressedDigis:HFNose'),
```
What happens if we run on a sample that doesn't contain the nose digis?

mariadalfonso replied:
tested successfully with runtheMatrix on scenario

22034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D46_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D46+RecoFullGlobal_2026D46+HARVESTFullGlobal_2026D46

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerCellCalibration.cc:
```diff
@@ -1,18 +1,25 @@
+#include "DataFormats/ForwardDetId/interface/HGCalTriggerDetId.h"
 
-//class constructor
+// class constructor
```
No reason to change this

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerCellCalibration.cc:
```diff
@@ -31,8 +38,17 @@ void HGCalTriggerCellCalibration::calibrateInMipT(l1t::HGCalTriggerCell& trgCell
-    trgCellMipP /= fCperMIP_;
+
+  if (DetId(trgCell.detId()).det() == DetId::HGCalTrigger and
+      HGCalTriggerDetId(trgCell.detId()).subdet() == HGCalTriggerSubdetector::HFNoseTrigger) {
```
If this kind of check is done in several places, it might be useful to add a function in the trigger tools class

mariadalfonso replied:
added the function in the triggertool

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerCellCalibration.cc:
```diff
@@ -31,8 +38,17 @@ void HGCalTriggerCellCalibration::calibrateInMipT(l1t::HGCalTriggerCell& trgCell
+    }
+
+  } else {
+    if (isSilicon && fCperMIP_ > 0) {
```
Can be `else if` in one single statement

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerCellCalibration.cc:
```diff
@@ -44,25 +60,52 @@ void HGCalTriggerCellCalibration::calibrateInMipT(l1t::HGCalTriggerCell& trgCell
 
-  DetId trgdetid(trgCell.detId());
-  unsigned trgCellLayer = triggerTools_.layerWithOffset(trgdetid);
+  unsigned trgCellLayer = triggerTools_.layerWithOffset(trgCell.detId());
```
Really, I don't like very much all these changes between raw id and detid. The code was more compact and readable before.

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerCellCalibration.cc:
```diff
@@ -44,25 +60,52 @@ void HGCalTriggerCellCalibration::calibrateInMipT(l1t::HGCalTriggerCell& trgCell
-           "and not with calibration coefficients = 0\n";
-  }
+  if (DetId(trgCell.detId()).det() == DetId::HGCalTrigger and
+      HGCalTriggerDetId(trgCell.detId()).subdet() == HGCalTriggerSubdetector::HFNoseTrigger) {
```
Definitely it should be put in the trigger tools

mariadalfonso replied:
yes, done

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerCellCalibration.cc:
```diff
@@ -44,25 +60,52 @@ void HGCalTriggerCellCalibration::calibrateInMipT(l1t::HGCalTriggerCell& trgCell
+    /* correct for the cell-thickness */
+    if (triggerTools_.isSilicon(trgCell.detId()) && thickCorr_ > 0) {
+      trgCellEt /= thickCorr_;
+    }
```
There is unnecessary code duplication. Things could be grouped; only the weights and the thickness corrections need to be different.

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerGeometryGenericMapping.cc:
```diff
@@ -96,6 +96,7 @@ unsigned HGCalTriggerGeometryGenericMapping::getLinksInModule(const unsigned mod
 unsigned HGCalTriggerGeometryGenericMapping::getModuleSize(const unsigned module_id) const { return 1; }
 
 GlobalPoint HGCalTriggerGeometryGenericMapping::getTriggerCellPosition(const unsigned trigger_cell_det_id) const {
+  std::cout << "HGCalTriggerGeometryGenericMapping::getTriggerCellPosition::::::: " << trigger_cell_det_id << std::endl;
```
`cout` to be removed

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerGeometryGenericMapping.cc:
```diff
@@ -111,6 +112,7 @@ bool HGCalTriggerGeometryGenericMapping::disconnectedModule(const unsigned modul
 
 unsigned HGCalTriggerGeometryGenericMapping::lastTriggerLayer() const {
   return eeTopology().dddConstants().layers(true);
+  // HFNOSE to update
```
I think there is no need to implement HFNose here. This is an old geometry class.

mariadalfonso replied:
ok changes to this file removed

jbsauvan reviewed (commented):
Here is a first set of comments. I will send the rest in a second step.

jbsauvan commented:
@mariadalfonso I will cherry-pick the PR #28249 from Sunanda with the constructor fix and make a separate (backport) PR here.

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -101,14 +111,20 @@ unsigned HGCalTriggerTools::layers(DetId::Detector type) const {
     layer = HGCSiliconDetId(id).layer();
-  } else if (id.det() == DetId::HGCalTrigger) {
+  } else if (id.det() == DetId::HGCalTrigger && (id.subdetId() == HGCalTriggerSubdetector::HGCalEETrigger ||
+                                                 id.subdetId() == HGCalTriggerSubdetector::HGCalHSiTrigger)) {
```
`HGCalTriggerDetId::subdet()` should be used and not `DetId::subdetId()`

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -101,14 +111,20 @@ unsigned HGCalTriggerTools::layers(DetId::Detector type) const {
+                                                 id.subdetId() == HGCalTriggerSubdetector::HGCalHSiTrigger)) {
     layer = HGCalTriggerDetId(id).layer();
+  } else if (id.det() == DetId::HGCalTrigger &&
+             HGCalTriggerDetId(id.rawId()).subdet() == HGCalTriggerSubdetector::HFNoseTrigger) {
```
call to `rawId()` to be removed

mariadalfonso replied:
removed everywhere

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -117,6 +133,12 @@ unsigned HGCalTriggerTools::layer(const DetId& id) const {
+  // for HFNose no offset
+  if (id.det() == DetId::HGCalTrigger &&
+      HGCalTriggerDetId(id.rawId()).subdet() == HGCalTriggerSubdetector::HFNoseTrigger)
+    return l;
```
Would be better to keep one single return statement at the end.

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -125,17 +147,24 @@ unsigned HGCalTriggerTools::layerWithOffset(const DetId& id) const {
-  } else if (id.det() == DetId::HGCalTrigger) {
-    em = (HGCalTriggerDetId(id).subdet() == HGCalTriggerSubdetector::HGCalEETrigger);
+  } else if (id.det() == DetId::HGCalTrigger && id.subdetId() == HGCalTriggerSubdetector::HGCalEETrigger) {
+    em = ((HGCalTriggerDetId(id).subdet() == HGCalTriggerSubdetector::HGCalEETrigger));
```
This has to be rewritten. The equality is necessarily true in that case and calls to `DetId::subdetId()` are not correct.

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -146,22 +175,34 @@ bool HGCalTriggerTools::isSilicon(const DetId& id) const {
+             HGCalTriggerDetId(id.rawId()).subdet() == HGCalTriggerSubdetector::HFNoseTrigger) {
+    silicon = true;
+  } else if (id.det() == DetId::HGCalTrigger && (id.subdetId() == HGCalTriggerSubdetector::HGCalEETrigger ||
+                                                 id.subdetId() == HGCalTriggerSubdetector::HGCalHSiTrigger)) {
```
Calls to `DetId::subdetId()` are not correct

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -146,22 +175,34 @@ bool HGCalTriggerTools::isSilicon(const DetId& id) const {
+  } else if (id.det() == DetId::Forward && id.subdetId() == ForwardSubdetector::HFNose) {
+    zside = HFNoseTriggerDetId(id).zside();
+  } else if (id.det() == DetId::HGCalTrigger && (id.subdetId() == HGCalTriggerSubdetector::HGCalEETrigger ||
+                                                 id.subdetId() == HGCalTriggerSubdetector::HGCalHSiTrigger)) {
```
Calls to `DetId::subdetId()` are not correct

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -203,8 +244,14 @@ int HGCalTriggerTools::thicknessIndex(const DetId& id, bool tc) const {
   else if (det == DetId::HGCalEE || det == DetId::HGCalHSi) {
     thickness = HGCSiliconDetId(id).type();
-  } else if (det == DetId::HGCalTrigger) {
+  } else if (det == DetId::Forward && id.subdetId() == ForwardSubdetector::HFNose) {
```
It will never enter here because of the V8 geometry `if` before

mariadalfonso replied:
fixed

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -203,8 +244,14 @@ int HGCalTriggerTools::thicknessIndex(const DetId& id, bool tc) const {
+  } else if (det == DetId::Forward && id.subdetId() == ForwardSubdetector::HFNose) {
+    thickness = HFNoseDetId(id).type();
+  } else if (id.det() == DetId::HGCalTrigger && (id.subdetId() == HGCalTriggerSubdetector::HGCalEETrigger ||
+                                                 id.subdetId() == HGCalTriggerSubdetector::HGCalHSiTrigger)) {
```
Calls to `DetId::subdetId()` are not correct

mariadalfonso replied:
not a general rule
error: 'const class DetId' has no member named 'subdet'; did you mean 'subdetId'?

changing all the id.subdetId() --> HGCalTriggerDetId(id).subdet()
works even if longer

jbsauvan replied:
`HGCalTriggerDetId::subdet()` should indeed be used. 
`DetId::subdetId()` is not the same as `HGCalTriggerDetId::subdet()`, it doesn't look at the same bits in the detid (`kSubdetMask = 0x7` vs `kHGCalSubdetMask = 0x3`). `DetId::subdetId()` (3 bits) is returning the 2 HGCAL subdet bits and the zside bit.

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -250,6 +297,7 @@ float HGCalTriggerTools::getLayerZ(const unsigned& layerWithOffset) const {
     subdet = HcalSubdetector::HcalEndcap;
     offset = lastLayerFH();
   }
+  // FIXME: to do HFNose (not sure what this is for)
```
In your case `layerWithOffset` is simply the layer (no offset), and the offset is 0.

mariadalfonso replied:
there is no switch based on the detector type.
Since I have 8 layers (less than your HGCEE) only I will stay always with offset = 0.

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -268,6 +316,7 @@ float HGCalTriggerTools::getLayerZ(const int& subdet, const unsigned& layer) con
           HcalSubdetector::HcalEndcap, geom_->bhTopology().dddConstants()->getEtaRange(1).second, layer);
     }
   }
+  // FIXME: to do HFNose (not sure what this is for)
```
This is the layer `z` position. Should be obtained from the detector topology.

mariadalfonso replied:
fixed

jbsauvan commented on L1Trigger/L1THGCal/src/backend/HGCalClusteringDummyImpl.cc:
```diff
@@ -33,15 +33,18 @@ void HGCalClusteringDummyImpl::clusterizeDummy(const std::vector<edm::Ptr<l1t::H
 void HGCalClusteringDummyImpl::calibratePt(l1t::HGCalCluster& cluster) {
   double calibPt = 0.;
 
+  // HFnose to update
```
Since this can be used in the clustering, the HFnose case should probably be handled.

mariadalfonso replied:
added protection on !isNose 
for now there is no scale correction

jbsauvan commented on L1Trigger/L1THGCal/src/veryfrontend/HGCalVFESummationImpl.cc:
```diff
@@ -50,8 +50,10 @@ void HGCalVFESummationImpl::triggerCellSums(const HGCalTriggerGeometryBase& geom
     uint32_t tcid = geometry.getTriggerCellFromCell(cellid);
     payload.emplace(tcid, 0);  // do nothing if key exists already
 
+    bool isNose = (DetId(cellid).det() == DetId::Forward && DetId(cellid).subdetId() == ForwardSubdetector::HFNose);
```
Should go in the trigger tools

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc:
```diff
@@ -374,6 +393,10 @@ void HGCalTriggerGeomTesterV9Imp2::beginRun(const edm::Run& /*run*/, const edm::
   no_neighbors_ = !checkNeighborConsistency();
+
+  no_trigger_ = false;
+  no_neighbors_ = false;
```
To be removed

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc:
```diff
@@ -463,22 +498,35 @@ bool HGCalTriggerGeomTesterV9Imp2::checkMappingConsistency() {
             edm::LogProblem("BadTriggerCell")
-                << "Unknown detector type " << id.det() << " " << trig_id.subdet() << "\n";
-            edm::LogProblem("BadTriggerCell") << "Cell ID " << HGCSiliconDetId(cell) << "\n";
+                << "Unknown detector type " << id.det() << " " << id.subdetId() << " " << id.rawId() << "\n";
```
Not correct to use `DetId::subdetId()`

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc:
```diff
@@ -598,6 +660,21 @@ bool HGCalTriggerGeomTesterV9Imp2::checkNeighborConsistency() {
+      // Skip trigger cells in module 0
+      uint32_t trigger_cell = triggerGeometry_->getTriggerCellFromCell(id);
+      uint32_t module = triggerGeometry_->getModuleFromTriggerCell(trigger_cell);
+      if (HGCalDetId(module).wafer() == 0)
```
It doesn't apply in this case since the Nose modules are not given as `HGCalDetId` and no modules are disconnected. A wafer==0 is meaning that this module is disconnected.

mariadalfonso replied:
thanks for the explanation, indeed this check is not needed

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc:
```diff
@@ -767,24 +887,36 @@ void HGCalTriggerGeomTesterV9Imp2::fillTriggerGeometry()
       triggerCellIEta_ = id_sc.ietaAbs();
       triggerCellIPhi_ = id_sc.iphi();
+      //    } else if (id.det()==DetId::Forward &&
+      //    id.subdetId()==ForwardSubdetector::HFNose)  {
```
Commented code to be removed

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc:
```diff
@@ -831,14 +980,20 @@ void HGCalTriggerGeomTesterV9Imp2::fillTriggerGeometry()
 
-    treeTriggerCells_->Fill();
+    // TEMPORARY
+    //    treeTriggerCells_->Fill();
```
To be uncommented

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc:
```diff
@@ -831,14 +980,20 @@ void HGCalTriggerGeomTesterV9Imp2::fillTriggerGeometry()
   // Loop over modules
   edm::LogPrint("TreeFilling") << "Filling modules tree";
+
+  std::cout << " looping in modules " << std::endl;
```
This `cout` doesn't bring much information

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc:
```diff
@@ -854,6 +1009,14 @@ void HGCalTriggerGeomTesterV9Imp2::fillTriggerGeometry()
       moduleIPhi_ = sc_id.iphi();
       module_ = 0;
+    } else if (id.det() == DetId::Forward and id.subdetId() == ForwardSubdetector::HFNose) {
+      HGCalDetId nose_id(module_triggercells.first);
```
This is not correct for Nose modules, right?

mariadalfonso replied:
fixed, HFNoseTriggerDetId nose_id(module_triggercells.first);

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc:
```diff
@@ -854,6 +1009,14 @@ void HGCalTriggerGeomTesterV9Imp2::fillTriggerGeometry()
+      moduleLayer_ = nose_id.layer();
+      moduleIEta_ = 0;
+      moduleIPhi_ = 0;
+      module_ = nose_id.wafer();
```
This is not correct for Nose modules, right?

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc:
```diff
@@ -944,7 +1133,8 @@ void HGCalTriggerGeomTesterV9Imp2::fillTriggerGeometry()
     //
-    treeModules_->Fill();
+    // TEMPORARY
+    //    treeModules_->Fill();
```
Code to be uncommented

jbsauvan commented on L1Trigger/L1THGCal/test/testHGCalL1TGeometryV9Imp2_Nose_cfg.py:
```diff
@@ -0,0 +1,155 @@
+process.hgcaltriggergeomtester = cms.EDAnalyzer(
+    "HGCalTriggerGeomTesterV9Imp2"
+    )
+process.test_step = cms.Path(process.hgcaltriggergeomtester)
```
It doesn't seem to be used. So this file doesn't seem to be testing the geometry. So whether the file name should be changed or the geometry testing modules should be used

mariadalfonso replied:
this file can test both geometry (line144) and create the PT collections ( line 147)

jbsauvan replied:
Could it be separated in 2 different configs (it is currently done like that for the other test files)? There are test files dedicated to purely geometry testing and test files dedicated to test the TPG chain.

mariadalfonso replied:
for now its' just me, for this PR I prefer have a single cfg.py to test both. Once 11_0 is out we can expand (i.e. add geometry V1, add multiple cfg, add dedicated DQM for jenkin tests ...)

jbsauvan replied:
Ok for now

mariadalfonso commented:
@jbsauvan 
i implemented most of your comments;
--> still in my list is to fine tune the module section of the L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc 
--> some will no do (spaces, reordering) + some is outside of this PR ( implementation of the missing things as links, non default geometry as V1 )

Have a look now and let me know if you have further comments.

This branch needs cms-sw#28249 that is already merged in the master of cms-sw.

jbsauvan commented:
@mariadalfonso my comment concerning the `V9Imp1` was not to implement anything. It was rather to throw an exception if the HFNose-version of the initialization method is called. Currently it runs silently. If something is not implemented but the user is using it, it should throw an exception.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend/HGCalHistoSeedingImpl.h:
```diff
@@ -147,7 +147,8 @@ class HGCalHistoSeedingImpl {
   SeedingPosition seedingPosition_;
 
-  unsigned nBins1_ = 42;
+  //  unsigned nBins1_ = 42;
```
Commented code to be removed

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/interface/backend/HGCalHistoSeedingImpl.h:
```diff
@@ -159,7 +160,10 @@ class HGCalHistoSeedingImpl {
   static constexpr unsigned neighbour_weights_size_ = 9;
-  static constexpr double kROverZMin_ = 0.076;
+  // HGCAL only value
+  //  static constexpr double kROverZMin_ = 0.076;
```
Commented code to be removed

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -184,7 +238,20 @@ unsigned HGCalTriggerGeometryV9Imp2::getModuleFromTriggerCell(const unsigned tri
     int ieta = ((trigger_cell_sc_id.ietaAbs() - ietamin_tc) / hSc_module_size_ + 1) * zside;
     int iphi = (trigger_cell_sc_id.iphi() - 1) / hSc_module_size_ + 1;
-    module_id = HGCScintillatorDetId(tc_type, layer, ieta, iphi);
+    module_id = HGCScintillatorDetId(tc_type, layer, ieta, iphi).rawId();
```
`rawId()` not needed

mariadalfonso replied:
the function 
unsigned HGCalTriggerGeometryV9Imp2::getModuleFromTriggerCell(
return an unsigned, so it's correct to have rawId()

jbsauvan replied:
It's correct, but an implicit conversion was done before and was also correct (and more compact). Please don't change existing code for no reason.

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -184,7 +238,20 @@ unsigned HGCalTriggerGeometryV9Imp2::getModuleFromTriggerCell(const unsigned tri
+    int waferv = trigger_cell_trig_id.waferV();
+
+    HFNoseDetIdToModule hfn;
+    module_id = hfn.getModule(HFNoseDetId(zside, layer, tc_type, waferu, waferv, 0, 0)).rawId();
```
Is `rawId()` needed?

mariadalfonso replied:
the function 
unsigned HGCalTriggerGeometryV9Imp2::getModuleFromTriggerCell(
return an unsigned, so it's correct to have rawId()

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -231,6 +299,22 @@ HGCalTriggerGeometryBase::geom_set HGCalTriggerGeometryV9Imp2::getCellsFromTrigg
+    std::vector<int> cellvs = trigger_cell_nose_id.cellV();
+    for (unsigned ic = 0; ic < cellus.size(); ic++) {
+      HFNoseDetId cell_det_id(zside, type, layer, waferu, waferv, cellus[ic], cellvs[ic]);
+      cell_det_ids.emplace(cell_det_id.rawId());
```
Is `rawId()` needed?

mariadalfonso replied:
not needed, fixed
fixed also in line 333  for HGC

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -295,6 +379,16 @@ HGCalTriggerGeometryBase::geom_set HGCalTriggerGeometryV9Imp2::getTriggerCellsFr
+    std::vector<HFNoseTriggerDetId> ids = hfn.getTriggerDetIds(module_nose_id);
+    for (auto const& idx : ids) {
+      if (validTriggerCellFromCells(idx))
+        trigger_cell_det_ids.emplace(idx.rawId());
```
Is `rawId()` needed?

mariadalfonso replied:
not needed in line 389, fixed in line 388
fixed also in line 418  for HGC

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -348,6 +443,16 @@ HGCalTriggerGeometryBase::geom_ordered_set HGCalTriggerGeometryV9Imp2::getOrdere
+    std::vector<HFNoseTriggerDetId> ids = hfn.getTriggerDetIds(module_nose_id);
+    for (auto const& idx : ids) {
+      if (validTriggerCellFromCells(idx))
+        trigger_cell_det_ids.emplace(idx.rawId());
```
Is `rawId()` needed?

mariadalfonso replied:
not needed in line 453, fixed in line 452
fixed also in line 481 for HGC

jbsauvan commented on L1Trigger/L1THGCal/python/hgcalBackEndLayer2Producer_cfi.py:
```diff
@@ -17,7 +17,8 @@
                       5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,  # 13 - 27
-                      3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3  # 28 - 41
+                      3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,  # 28 - 41
+                      3, 3, 3, 3 # 42 - 45
```
Should actually be the other way round. The 4 bins to be added are at low R/Z values. So four `13` (although the optimal smoothing size is maybe different) should be added at the beginning.

mariadalfonso replied:
fixed, thanks for the explanation

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerCellCalibration.cc:
```diff
@@ -44,25 +56,53 @@ void HGCalTriggerCellCalibration::calibrateInMipT(l1t::HGCalTriggerCell& trgCell
+    /* correct for the cell-thickness */
+    if (triggerTools_.isSilicon(trgdetid) && thickCorr_ > 0) {
+      trgCellEt /= thickCorr_;
+    }
```
Would be good to reduce code duplication. Only the weight and corr values are different between hgcal and nose

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -49,10 +53,12 @@ void HGCalTriggerTools::eventSetup(const edm::EventSetup& es) {
+    // not think to anything better to distinguish a TC detId
   }
 
+  // FIXME: to do HFNose
```
In principle it is ok, since it delegates the calculation to the geometry which correctly computes the nose TC positions

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -117,6 +133,11 @@ unsigned HGCalTriggerTools::layer(const DetId& id) const {
   unsigned int l = layer(id);
+
+  // for HFNose no offset
+  if (id.det() == DetId::HGCalTrigger && HGCalTriggerDetId(id).subdet() == HGCalTriggerSubdetector::HFNoseTrigger)
```
This method could be called also for standard cells. So the `if` should include that case. The clearest would probably to have
```cpp
if(isNose(id))
...
```

mariadalfonso replied:
done

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerTools.cc:
```diff
@@ -250,6 +305,7 @@ float HGCalTriggerTools::getLayerZ(const unsigned& layerWithOffset) const {
     offset = lastLayerFH();
   }
+  // note for HFnose offset is always zero since we have less layers than HGCEE
   return getLayerZ(subdet, layerWithOffset - offset);
```
There will be an issue with the subdet value. Not sure how to deal with that, though. Except if we simply remove this method to prevent an incorrect usage of it.

jbsauvan commented on L1Trigger/L1THGCal/src/backend/HGCalShowerShape.cc:
```diff
@@ -87,6 +87,8 @@ int HGCalShowerShape::coreShowerLength(const l1t::HGCalMulticluster& c3d,
     if (!pass(*id_cluster.second, c3d))
       continue;
     unsigned layer = triggerGeometry.triggerLayer(id_cluster.second->detId());
+    if (triggerTools_.isNose(id_cluster.second->detId()))
+      nlayers = triggerTools_.layers(ForwardSubdetector::HFNose);
```
Is it possible to have a cluster reconstructed partly in HGCAL and partly in the nose? What would happen in that case?

mariadalfonso replied:
the possible overlap is around eta 3
the idea is to have this done at correlator level and not at the calo level.
So we keep separated the FW and SW for HFnose and HGC.
Once the 11_0 is out the overlap region can be better studied.

jbsauvan replied:
Right now  `nLayers`, which should be constant for a given cluster, is modified when a Nose TC is found in the cluster. Which means it will break the output of this function when a cluster is both in the HGCAL and the nose. This should be fixed.

mariadalfonso replied:
fixed by putting back the original default is noNose

jbsauvan replied:
Now I think the `nlayers` value is correct. Still to be fixed in the other method `percentileLayer`

mariadalfonso replied:
fixed the percentileLayer as well

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc:
```diff
@@ -854,6 +996,15 @@ void HGCalTriggerGeomTesterV9Imp2::fillTriggerGeometry()
+      moduleIEta_ = 0;
+      moduleIPhi_ = 0;
+      // FIXME
+      //      module_ = nose_id.wafer();
```
If it is not defined, should be put to 0

mariadalfonso replied:
ok, set to 0

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp2.cc:
```diff
@@ -944,7 +1121,8 @@ void HGCalTriggerGeomTesterV9Imp2::fillTriggerGeometry()
     //
-    treeModules_->Fill();
+    // TEMPORARY
+    //    treeModules_->Fill();
```
Code to be uncommented

mariadalfonso replied:
done

mariadalfonso commented:
validation plot at this stage are here 
http://dalfonso.web.cern.ch/dalfonso/HGCAL/L1Plot/afterPR/

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp1.cc:
```diff
@@ -171,8 +173,42 @@ void HGCalTriggerGeometryV9Imp1::initialize(const HGCalGeometry* hgc_ee_geometry
+  fillNeighborMap(l1tCellNeighborsMapping_, trigger_cell_neighbors_,
+                  false);  // silicon
+  fillNeighborMap(l1tCellNeighborsSciMapping_, trigger_cell_neighbors_sci_,
+                  true);  // scintillator
```
Now that there is an exception thrown just above, there is no need to have this code

mariadalfonso replied:
removed

mariadalfonso commented:
@jbsauvan 
Fixed this morning comments. What is left ?

mariadalfonso commented:
@jbsauvan done

mariadalfonso commented:
which kind of validation you do on this PR ? and in general ?
--> in principle would be good to factorize out the effect of this PR (in principle none) with the rest of the changes you guys made

jbsauvan commented:
Thanks. 
For the validation, we are comparing various distributions for TCs and clusters before and after the updates. I will run it on a sample without the nose, which shouldn't show any difference.
Then do you know if there are some relval samples with the nose included?

mariadalfonso commented:
We have two workflows for the nose.
21234.0 -- D44 (here you have the HGC a la TDR)
22434.0 -- D47 (here you have the HGC a la D41 with V11)

21234.0 usually run in the PR integrations in central cms-sw

jbsauvan commented:
Has it been tested with the older D35 geometry (which was used for the MTD samples)?  In that case it tries to find the HFNose geometry, while it should not.
```
----- Begin Fatal Exception 29-Oct-2019 09:18:37 CET-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  stream begin Run run: 1 stream: 0
   [1] Calling method for module HGCalVFEProducer/'hgcalVFEProducer'
   [2] Using EventSetup component HGCalTriggerGeometryESProducer/'hgcalTriggerGeometryESProducer' to make data HGCalTriggerGeometryBase/'' in record CaloGeometryRecord                                                                                            
Exception Message:
No data of type "HGCalGeometry" with label "HGCalHFNoseSensitive" in record "IdealGeometryRecord"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------
```

mariadalfonso commented:
I usually tested on this and works (note HFnose is not there)
22034.0 D46 Phase2C9_timing_layer_bar

It fail for older ones, and I will look into it
20434.0 D41 Phase2C8_timing_layer_bar
20034.0 D35 Phase2C4_timing_layer_bar

mariadalfonso commented:
@jbsauvan 
I fixed the old workflows, now it runs for the D46, D41 and D35. 
(Note HFnose is only in the D44 and D47)

(you can inspect changes here https://github.com/cms-sw/cmssw/commit/1fa12a8f17886b80f122288dea0b31fb02302679)

mariadalfonso commented:
rolled back some of the useless formatting changes.

jbsauvan commented:
Thanks @mariadalfonso 
It seems the change in seeding binning has some significant impact on the produced clusters:
https://jsauvan.web.cern.ch/jsauvan/private/hgcal/validation/pr400/
It's not clear to me why the impact is so big. We will need to perform some study.
But for the moment could we revert the seeding parameters back to their original values, and add a custom for HFNose eras? (This would mean that `kROverZMin_` and `...Max_` should actually be configurable)

mariadalfonso commented:
@jbsauvan  
it's not trivial to customize with era.
We have the Eras.Modifier_phase2_hfnose_cff but to not mix the HGC and HFnose in the D46 scenario I need to duplicate all the parameters and then find a way to switch inside the code based on the volume.
I will try to understand tomorrow/week-end why the simple extension of the binning didn't worked.

mariadalfonso commented:
@jbsauvan 

To keep the old binning for HGCAL, I have fixed it adding one extra bin one 42-->43, and extending the kROverZMin_  such that the boundary for R1 and R2 fall in the same position as w/o the HFNOSE.

On a separate note, I'm importing now the whole vectors of fcPerMip and thickness correction and correct on the fly based based on the thickness (previously I was only importing the 120 um).

jbsauvan commented:
@mariadalfonso, the differences are bigger than before ([Validation 2](https://jsauvan.web.cern.ch/jsauvan/private/hgcal/validation/pr400_2/)).
And I'm not sure to understand the changes, the R/Z bin values can now be negative (and with large values), which is nonphysical.
Is it really not possible to switch from one binning to an other based on the HFNose era modifier? The only thing I would like to ensure is that when the nose is not there the results stay the same.

mariadalfonso commented:
@jbsauvan 

I updated the branch restored the original bondaries/old bins.
I added a customization for the hfnose era.
In this situation D35,D41,D46 should be ok.
However in the D44 and D47 , scenario where both HGCAL and HFnose the HGCAL, the binning/boundary of the seeding class would be changed for both HGCAL and HFnose.

We can test one more time, at least we will know if you see other residual differences; but in long term we need to fix this seeding parameters

jbsauvan commented on L1Trigger/L1THGCal/interface/backend/HGCalHistoSeedingImpl.h:
```diff
@@ -159,8 +159,9 @@ class HGCalHistoSeedingImpl {
-  static constexpr double kROverZMin_ = 0.076;
-  static constexpr double kROverZMax_ = 0.58;
+  double kROverZMin_ = 0.076;
+  double kROverZMax_ = 0.58;
```
Better to define these as `const` since they should be constant after construction, although not constant at compile time as before.

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerCellCalibration.cc:
```diff
@@ -1,24 +1,32 @@
+      thickCorrNose_(conf.getParameter<std::vector<double>>("thickCorrNose")),
+      dEdX_weights_(conf.getParameter<std::vector<double>>("dEdXweights")),
+      dEdX_weights_Nose_(conf.getParameter<std::vector<double>>("dEdXweightsNose")) {
+  if (fCperMIP_ <= 0 or fCperMIPnose_[0] <= 0) {
```
The vector size should be checked. Or at least `.at()` should be used instead of `[]` such that  out-of-bound check is performed

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerCellCalibration.cc:
```diff
@@ -1,24 +1,32 @@
                                          "applied to correct trigger cell energies.";
   }
-  if (thickCorr_ <= 0) {
+  if (thickCorr_ <= 0 or thickCorrNose_[0] <= 0) {
```
ditto

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerCellCalibration.cc:
```diff
@@ -1,24 +1,32 @@
+  DetId trgdetid(trgCell.detId());
+  bool isSilicon = triggerTools_.isSilicon(trgdetid);
+  bool isNose = triggerTools_.isNose(trgdetid);
+  int thickness_ = isNose ? triggerTools_.thicknessIndex(trgdetid, false) : 0;
```
It is a local variable. Trailing `_` are reserved for class members

jbsauvan replied:
Is there a reason to force to `0` in the case it is not a nose cell? If the thickness is not used for non-nose cells, it is clearer to keep the real thickness value in all cases

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerCellCalibration.cc:
```diff
@@ -44,25 +55,44 @@ void HGCalTriggerCellCalibration::calibrateInMipT(l1t::HGCalTriggerCell& trgCell
   DetId trgdetid(trgCell.detId());
+  bool isNose = triggerTools_.isNose(trgdetid);
   unsigned trgCellLayer = triggerTools_.layerWithOffset(trgdetid);
+  int thickness_ = isNose ? triggerTools_.thicknessIndex(trgdetid, false) : 0;
```
Same as above

jbsauvan commented:
Validation plots are now fine in the case with no HFNose ([Validation 3](https://jsauvan.web.cern.ch/jsauvan/private/hgcal/validation/pr400_3/))

mariadalfonso commented:
@jbsauvan comments resolved.
